### PR TITLE
PLAT-2040 Implement modules

### PR DESCRIPTION
**Pull request by ian-axelrod**
Adds main module, lambda-function module, and service-autoscaling module.

PLAT-2040

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  tag:
+    machine: true
+    parameters:
+      version:
+          description: "Module version"
+          type: string
+    steps:
+      - add_ssh_keys:
+          fingerprints:
+            - "bd:5a:bf:66:f2:6a:cf:d4:45:32:4b:19:1a:db:87:2d"
```
Comment (mbrochstein) [inline .circleci/config.yml:41]: Why did you need to include this?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+.terraform/
```
Comment (mbrochstein) [inline .gitignore]: Look at https://github.com/github/gitignore/blob/master/Terraform.gitignore

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,45 @@
+terraform {
+  required_version = ">= 0.11.11"
```
Comment (mbrochstein) [inline main.tf:2]: Why is this the pinned version? You're building it with TF 0.12.2, but requiring a 0.11.x version. Seems like this should be lower, or higher.

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  name                = "${data.aws_lambda_function.compute_queue_backlog.function_name}-${var.service_name}"
+  description         = "Schedule execution of ${data.aws_lambda_function.compute_queue_backlog.function_name} to compute queue backlog metrics for ${var.service_name} ${var.queue_name} queue."
+  schedule_expression = "${var.lambda_invocation_interval}"
+}
+
+resource "aws_cloudwatch_event_target" "compute_queue_backlog" {
+  rule      = "${aws_cloudwatch_event_rule.compute_queue_backlog.name}"
+  target_id = "${data.aws_lambda_function.compute_queue_backlog.function_name}-${var.service_name}"
+  arn       = "${data.aws_lambda_function.compute_queue_backlog.arn}"
+  input     = "{\"cluster_name\": \"${var.cluster_name}\", \"service_name\": \"${var.service_name}\", \"queue_name\": \"${var.queue_name}\", \"queue_owner_aws_account_id\": \"${var.queue_owner_aws_account_id != "" ? var.queue_owner_aws_account_id : data.aws_caller_identity.current.account_id}\", \"est_msgs_per_sec\": \"${var.service_est_msgs_per_sec}\"}"
```
Comment (mbrochstein) [inline main.tf]: Best practice would be to put this JSON in a template file and load it with a data module. Inlining escaped JSON is difficult to modify.

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+resource "aws_lambda_permission" "compute_queue_backlog" {
+  statement_id  = "AllowExecutionFromCloudWatch"
+  action        = "lambda:InvokeFunction"
+  function_name = "${data.aws_lambda_function.compute_queue_backlog.function_name}"
+  principal     = "events.amazonaws.com"
+  source_arn    = "${aws_cloudwatch_event_rule.compute_queue_backlog.arn}"
+}
+
+data "aws_caller_identity" "current" {}
```
Comment (mbrochstein) [inline main.tf:57]: What if the queue owner isn't the same as the executing account? This is the case with felix.

File: modules/lambda-function/main.tf [hunk 6]
```diff
@@ -0,0 +1,82 @@
+terraform {
+  required_version = ">= 0.11.11"
```
Comment (mbrochstein) [inline modules/lambda-function/main.tf:2]: Make sure this is the correct version - same comment as above.

File: modules/lambda-function/variables.tf [hunk 7]
```diff
@@ -0,0 +1,8 @@
+variable "name" {
+  description = "The lambda function name."
+}
+
+variable "execution_role_arn" {
+  description = "IAM role arn to use for lambda execution. If not supplied, this module will create a role with necessary permissions."
```
Comment (mbrochstein) [inline modules/lambda-function/variables.tf]: You may want to include the required permissions here.

File: modules/lambda-function/variables.tf [hunk 8]
```diff
@@ -0,0 +1,8 @@
+variable "name" {
+  description = "The lambda function name."
```
Comment (mbrochstein) [inline modules/lambda-function/variables.tf:2]: No default?

File: modules/lambda-function/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  value       = "${aws_lambda_function.compute_queue_backlog.arn}"
+  description = "This is compute queue backlog lambda's unqualified arn."
+}
+
+output "name" {
+  value       = "${aws_lambda_function.compute_queue_backlog.function_name}"
+  description = "This is compute queue backlog lambda's function name."
+}
+
+output "execution_role_arn" {
```
Comment (mbrochstein) [inline modules/lambda-function/outputs.tf:11]: If you pass in an execution role and one isn't created, this will be empty or error (not sure which)

File: modules/service-autoscaling/main.tf [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        name  = "QueueName"
+        value = "${var.queue_name}"
+      }
+    }
+  }
+}
+
+resource "aws_appautoscaling_policy" "queue_requires_consumer_scaling_policy" {
+  name               = "${var.service_name}-${var.queue_name}-queue-requires-consumer-scaling-policy"
+  policy_type        = "StepScaling"
```
Comment (mbrochstein) [inline modules/service-autoscaling/main.tf:50]: Is there a reason to make this a variable as well with a default to StepScaling?

File: modules/service-autoscaling/variables.tf [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  description = "ECS cluster name. Uses default cluster if not supplied."
+  default     = ""
+}
+
+variable "service_name" {
+  description = "ECS service name"
+}
+
+variable "max_capacity" {
+  description = "Maximum number of tasks that the autoscaling policy can set for the service."
```
Comment (mbrochstein) [inline modules/service-autoscaling/variables.tf:11]: This should be set to a reasonable default

File: modules/service-autoscaling/variables.tf [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  description = "Minimum number of tasks that the autoscaling policy can set for the service."
+  default     = 0
+}
+
+variable "queue_name" {
+  description = "Name of the queue to compute QueueBacklog metric for."
+}
+
+variable "target_value" {
+  description = "Queue backlog (in seconds) to maintain for the service when under maximum load."
```
Comment (mbrochstein) [inline modules/service-autoscaling/variables.tf]: Is there a good reasonable default to set this to?

**Review by mbrochstein — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 1 above)
Comment (ian-axelrod) [reply .circleci/config.yml:41]: I implemented auto-tagging via CircleCI, since Terraform Registry finds new versions of a module via git annotated tags that conform to semver. The deploy key that CircleCI uses is read-only, so I could not push the annotated tag. I created my own read/write deploy key instead, as otherwise I'd need to create a machine user or use one of my personal keys (no thnx) with significant access to our private repositories. The deploy key grants r/w access to this repository only. I needed to add the fingerprint so CircleCi will pick up the key and use it.

File: main.tf (hunk 3 above)
Comment (ian-axelrod) [reply main.tf:2]: I'm only validating. I want this module to be compatible with Terraform 0.12.x and 0.11.x, so that I do not have to make any changes when we upgrade Terraform in other places. I think the best solution would be to validate both 0.11.x and 0.12.x. I'll do that unless you have objections.

File: main.tf (hunk 5 above)
Comment (ian-axelrod) [reply main.tf:57]: Systems that need to grant the lambda access to a queue in a different account will need to add the permissions, like we do for felix. I could add the permissions, like I do for cloudwatch rules; however, the situation is different because the queue resource is not something this module creates, so I do not want to start adding policy documents to resources that are largely unknown to the module. Thoughts?

File: modules/lambda-function/variables.tf (hunk 8 above)
Comment (ian-axelrod) [reply modules/lambda-function/variables.tf:2]: I removed it. I guess I could use a default. I figured if someone is using only this module they have a more advanced use case, so I moved the default up one level to the parent module. Hate re-declaring defaults, heh.

File: modules/lambda-function/outputs.tf (hunk 9 above)
Comment (ian-axelrod) [reply modules/lambda-function/outputs.tf:11]: I mention this in the description. It will not error. It will be empty, which is by design.

File: modules/service-autoscaling/variables.tf (hunk 12 above)
Comment (ian-axelrod) [reply modules/service-autoscaling/variables.tf]: Not in my opinion. It's too dependent on the service. There's probably reasonable upper and lower limits (in most cases), but the range within those is too large to really give a default.

File: main.tf (hunk 3 above)
Comment (ian-axelrod) [reply main.tf:2]: ping @mbrochstein How do you feel about my approach?

File: modules/service-autoscaling/variables.tf (hunk 11 above)
Comment (ian-axelrod) [reply modules/service-autoscaling/variables.tf:11]: The only reasonable default is `1`, in my opinion. I'll set it to that.

File: modules/service-autoscaling/main.tf (hunk 10 above)
Comment (ian-axelrod) [reply modules/service-autoscaling/main.tf:50]: No, since only StepScaling and TargetTrackingScaling policies are the only valid ECS autoscaling policies. TargetTrackingScaling doesn't make sense for this policy. I can update this if that ever changes, though.

File: modules/service-autoscaling/variables.tf (hunk 11 above)
Comment (ian-axelrod) [reply modules/service-autoscaling/variables.tf:11]: Done

File: main.tf (hunk 4 above)
Comment (ian-axelrod) [reply main.tf]: Done

File: .gitignore (hunk 2 above)
Comment (ian-axelrod) [reply .gitignore]: Updated

**Review by Shikkic — APPROVED** [review]
I unfortunately don't think I know enough to validate most of this. It looks fine from what I can tell and I don't think I have any feedback, but I wouldn't lean on that too much. I think @mbrochstein review will be more helpful. I just didn't want to block this with a pending review from me.

File: main.tf (hunk 3 above)
Comment (mbrochstein) [reply main.tf:2]: Pinning the versions makes sense, but the module can't be compatible with both tf 0.11.x and tf 0.12.x because of the changes to HCL, right?

File: main.tf (hunk 5 above)
Comment (mbrochstein) [reply main.tf:57]: I agree - trying to think about the best way to handle this that doesn't involve creating permissions that won't be used. Maybe having a variable for `account_id` and using the current calling identity if none is specified?

File: modules/lambda-function/variables.tf (hunk 8 above)
Comment (mbrochstein) [reply modules/lambda-function/variables.tf:2]: Assume that all modules MAY be used independently.

File: modules/service-autoscaling/variables.tf (hunk 11 above)
Comment (mbrochstein) [reply modules/service-autoscaling/variables.tf:11]: That makes sense. The idea being easy as possible to implement and get running with lots of configurations. README should point out that the default is 1, which isn't really useful.

File: modules/service-autoscaling/variables.tf (hunk 12 above)
Comment (mbrochstein) [reply modules/service-autoscaling/variables.tf]: I think it's probably worth setting something insanely conservative as the default, mostly to simplify the ability to get it up and running as soon as possible.

**Review by mbrochstein — CHANGES_REQUESTED** [review]

File: main.tf (hunk 3 above)
Comment (ian-axelrod) [reply main.tf:2]: From what I can tell, this module is compatible with both. I test on 0.12.2 and the module has already been used on 0.11.x through other repositories that pull it in. I made an effort uses of 0.12.x-only features / syntax.

File: main.tf (hunk 5 above)
Comment (ian-axelrod) [reply main.tf:57]: That's what I do. `queue_owner_aws_account_id`

File: main.tf (hunk 3 above)
Comment (ian-axelrod) [reply main.tf:2]: To extend upon this, I do not want to pin a module to a specific terraform version. This module is public, so I have designed it under the assumption I cannot  (well, should not) control the exact version the module requires. I would like to avoid upper-bound version constraints until it's demonstrated that there is an incompatibility with some version.

File: modules/lambda-function/variables.tf (hunk 8 above)
Comment (ian-axelrod) [reply modules/lambda-function/variables.tf:2]: Done

File: modules/lambda-function/variables.tf (hunk 7 above)
Comment (ian-axelrod) [reply modules/lambda-function/variables.tf]: Done

File: modules/service-autoscaling/variables.tf (hunk 12 above)
Comment (ian-axelrod) [reply modules/service-autoscaling/variables.tf]: Ok. Done.

File: main.tf (hunk 5 above)
Comment (mbrochstein) [reply main.tf:57]: Got it  - missed that.

**Review by mbrochstein — APPROVED** [review]
